File: ios/Modules/Sources/Calls/impl/HttpIceConfigProvider.swift
import Foundation
import Net

/// STUN / TURN servers fetched from the node's `/v1/calls/ice`.
/// List changes when the operator rotates TURN creds, so [CallController]
/// refreshes on every new call instead of caching forever.
public final class HttpIceConfigProvider: IceConfigProvider {
    private let http: HttpClient

    public init(http: HttpClient) { self.http = http }

    public func current() async -> [IceServer] {
        do {
            let resp = try await http.send(.get("v1/calls/ice"), IceResp.self)
            return resp.servers.map {
                IceServer(urls: $0.urls, username: $0.username, credential: $0.credential)
            }
        } catch {
            // Fall back to Google's public STUN so a fresh install can at
            // least complete peer discovery on unrestricted networks.
            return [IceServer(urls: ["stun:stun.l.google.com:19302"])]
        }
    }

    private struct IceDto: Decodable {
        let urls: [String]
        let username: String?
        let credential: String?
    }
    private struct IceResp: Decodable { let servers: [IceDto] }
}

File: android/app/src/main/java/sol/vortexx/android/calls/impl/HttpIceConfigProvider.kt
package sol.vortexx.android.calls.impl

import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.http.isSuccess
import kotlinx.serialization.Serializable
import sol.vortexx.android.calls.api.IceConfigProvider
import sol.vortexx.android.calls.api.IceServer
import sol.vortexx.android.net.impl.VortexHttpClient
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class HttpIceConfigProvider @Inject constructor(
    private val http: VortexHttpClient,
) : IceConfigProvider {

    override suspend fun current(): List<IceServer> = runCatching {
        val resp = http.client.get("v1/calls/ice")
        if (!resp.status.isSuccess()) return@runCatching DEFAULT_ICE
        resp.body<IceResp>().servers.map {
            IceServer(urls = it.urls, username = it.username, credential = it.credential)
        }
    }.getOrDefault(DEFAULT_ICE)

    @Serializable private data class IceDto(
        val urls: List<String>,
        val username: String? = null,
        val credential: String? = null,
    )
    @Serializable private data class IceResp(val servers: List<IceDto>)

    private companion object {
        // Fallback when the node doesn't surface TURN creds (e.g. fresh
        // install). Google's public STUN works for unblocked LANs.
        val DEFAULT_ICE = listOf(IceServer(urls = listOf("stun:stun.l.google.com:19302")))
    }
}
